maze_dataset/tokenization/maze_tokenizer.py

```diff
@@ -388,3 +429,723 @@ def clear_cache(self):
+            - `path`: Tokens representing the path
+            """
+            pass
+
+    @serializable_dataclass(frozen=True, kw_only=True)
+    class AOTP(PromptSequencer):
+        def _sequence_tokens(
+            self,
+            adj_list: list[str],
+            origin: list[str],
+            target: list[str],
+            path: list[str],
+        ) -> list[str]:
+            return [
+                VOCAB.ADJLIST_START,
+                *adj_list,
+                VOCAB.ADJLIST_END,
+                VOCAB.ORIGIN_START,
+                *origin,
+                VOCAB.ORIGIN_END,
+                VOCAB.TARGET_START,
+                *target,
+                VOCAB.TARGET_END,
+                VOCAB.PATH_START,
+                *path,
+                VOCAB.PATH_END,
+            ]
+
+    @serializable_dataclass(frozen=True, kw_only=True)
+    class AOP(PromptSequencer):
+        include_target_special_tokens: bool = serializable_field(default=True)
+
+        def _sequence_tokens(
+            self,
+            adj_list: list[str],
+            origin: list[str],
+            target: list[str],
+            path: list[str],
+        ) -> list[str]:
+            return [
+                VOCAB.ADJLIST_START,
+                *adj_list,
+                VOCAB.ADJLIST_END,
+                VOCAB.ORIGIN_START,
+                *origin,
+                VOCAB.ORIGIN_END,
+                *(
+                    [VOCAB.TARGET_START, VOCAB.TARGET_END]
+                    if self.include_target_special_tokens
+                    else ()
+                ),
+                VOCAB.PATH_START,
+                *path,
+                VOCAB.PATH_END,
+            ]
+
+
+def _load_tokenizer_element(
+    data: dict[str, Any], namespace: type[_TokenizerElementNamespace]
+) -> TokenizerElement:
+    key: str = namespace.key
+    format: str = data[key]["__format__"]
+    cls_name: str = format.split("(")[0]
+    # return getattr(namespace, cls_name).deserialize(data[key])
+    cls: type[TokenizerElement] = getattr(namespace, cls_name)
+    # TODO: Refactor line below with `zanj.loading.load_item_recursive` to properly leverage zanj
+    return cls(**{k: v for k, v in data[key].items() if k != "__format__"})
+
+
+@serializable_dataclass(frozen=True, kw_only=True)
+class MazeTokenizer2(SerializableDataclass):
+    """Tokenizer for mazes
+
+    # Development
+    - To ensure backwards compatibility, the default constructor must always return a tokenizer equivalent to the legacy `TokenizationMode.AOTP_UT_Uniform`.
+    - Furthermore, the mapping reflected in `from_legacy` must also be maintained.
+    - Updates to `MazeTokenizer2` or the `TokenizerElement` hierarchy must maintain that behavior.
+    """
+
+    prompt_sequencer: PromptSequencers.PromptSequencer = serializable_field(
+        default=PromptSequencers.AOTP(),
+        # serialization_fn=lambda x: repr(x),
+        loading_fn=lambda x: _load_tokenizer_element(x, PromptSequencers),
```

Comment:
iirc the `loading_fn` actually passes the outer dict [1] and so this is what causes the complexity in `_load_tokenizer_element` -- maybe a way to refactor around this, but this would be a non-breaking change in the future and so should be pushed off


[1]: yes, this is horrible behavior and there is an open issue on `muutils` to add a different option here

Reply:
this is refactored in pr #37 , don't touch

Reply:
The refactor in PR #37 is not robust and currently broken for some new features in that PR. No need to touch anything in this PR, but it will some attention in that one.

Reply:
closing, moving to future PR